# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [95]:
import pandas as pd
url= 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv'
marketing_customer_analysis = pd.read_csv(url)

In [96]:
marketing_customer_analysis

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,...,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10905,10905,FE99816,Nevada,15563.369440,No,Premium,Bachelor,2011-01-19,Unemployed,F,...,7,Personal Auto,Personal L1,Offer3,Web,1214.400000,Luxury Car,Medsize,A,1
10906,10906,KX53892,Oregon,5259.444853,No,Basic,College,2011-01-06,Employed,F,...,6,Personal Auto,Personal L3,Offer2,Branch,273.018929,Four-Door Car,Medsize,A,1
10907,10907,TL39050,Arizona,23893.304100,No,Extended,Bachelor,2011-02-06,Employed,F,...,2,Corporate Auto,Corporate L3,Offer1,Web,381.306996,Luxury SUV,Medsize,A,2
10908,10908,WA60547,California,11971.977650,No,Premium,College,2011-02-13,Employed,F,...,6,Personal Auto,Personal L1,Offer1,Branch,618.288849,SUV,Medsize,A,2


1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [97]:
mask = (marketing_customer_analysis["total_claim_amount"] < 1000) & (marketing_customer_analysis["response"] == "Yes")
low_claim_amount = marketing_customer_analysis[mask]

In [98]:
low_claim_amount.sort_values(by = "total_claim_amount", ascending = False)

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
3338,3338,AM26717,California,3123.080994,Yes,Basic,Bachelor,2011-02-19,Unemployed,M,...,1,Personal Auto,Personal L1,Offer1,Agent,960.115399,SUV,Large,A,2
3558,3558,ST96866,Washington,3123.080994,Yes,Basic,Bachelor,2011-02-19,Unemployed,M,...,1,Special Auto,Special L3,Offer1,Agent,960.115399,SUV,Large,A,2
4414,4414,JW24212,Oregon,3123.080994,Yes,Basic,Bachelor,2011-02-19,Unemployed,M,...,1,Personal Auto,Personal L3,Offer1,Agent,960.115399,SUV,Large,A,2
7492,7492,YS74016,California,3123.080994,Yes,Basic,Bachelor,2011-02-19,Unemployed,M,...,1,Personal Auto,Personal L2,Offer1,Agent,960.115399,SUV,Large,A,2
8377,8377,NT76210,Nevada,3123.080994,Yes,Basic,Bachelor,2011-02-19,Unemployed,M,...,1,Personal Auto,Personal L3,Offer1,Agent,960.115399,SUV,Large,A,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2792,2792,CJ81799,Oregon,7462.926263,Yes,Extended,College,2011-01-28,Employed,F,...,5,Personal Auto,Personal L2,Offer2,Call Center,7.345946,Two-Door Car,Small,A,1
7886,7886,EZ70589,Washington,7462.926263,Yes,Extended,College,2011-01-28,Employed,F,...,5,Personal Auto,Personal L1,Offer2,Call Center,7.345946,Two-Door Car,Small,A,1
9131,9131,TR81766,California,7462.926263,Yes,Extended,College,2011-01-28,Employed,F,...,5,Corporate Auto,Corporate L2,Offer2,Call Center,7.345946,Two-Door Car,Small,A,1
2400,2400,GJ99406,Arizona,7462.926263,Yes,Extended,College,2011-01-28,Employed,F,...,5,Corporate Auto,Corporate L2,Offer2,Call Center,7.345946,Two-Door Car,Small,A,1


2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [99]:
marketing_customer_analysis.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type', 'month'],
      dtype='str')

In [100]:
grouped = marketing_customer_analysis.groupby('response')

In [101]:
yes = grouped.get_group("Yes")

In [102]:
yes.groupby(["policy_type", "gender"])[["monthly_premium_auto", "customer_lifetime_value", "total_claim_amount"]].mean().round(2)

monthly_premium_auto  customer_lifetime_value  \
policy_type    gender                                                  
Corporate Auto F                      94.30                  7712.63   
               M                      92.19                  7944.47   
Personal Auto  F                      99.00                  8339.79   
               M                      91.09                  7448.38   
Special Auto   F                      92.31                  7691.58   
               M                      86.34                  8247.09   

                       total_claim_amount  
policy_type    gender                      
Corporate Auto F                   433.74  
               M                   408.58  
Personal Auto  F                   452.97  
               M                   457.01  
Special Auto   F                   453.28  
               M                   429.53

Compare these insights to total_claim_amount patterns, and discuss which segments appear most profitable or low-risk for the company.

- Most Profitable: Personal Auto (Female) generates the highest monthly premium (98.99) and lifetime value (8,339.79), easily absorbing claims (452.97). 
- Lowest Risk: Corporate Auto (Male) delivers the lowest claim amount (408.58) alongside stable premium revenue (92.19).   
- Consistent: Corporate Auto (Female) similarly exhibits well-managed claim levels (433.74) with steady premiums (94.30).   
- Highest Risk: Personal Auto (Male) presents the narrowest margins, pairing peak claim severity (457.01) with the lowest customer lifetime value (7,448.38). 

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [108]:
policy_num = marketing_customer_analysis["number_of_policies"] > 0
filtered = marketing_customer_analysis[policy_num]
state_count = filtered.groupby("state")["customer"].count()

In [109]:
more_than500 = (state_count[state_count > 500])

In [110]:
more_than500

state
Arizona       1937
California    4183
Nevada         993
Oregon        2909
Washington     888
Name: customer, dtype: int64

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [113]:
marketing_customer_analysis.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median']).round(2)

max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

- Stable Medians: Typical customer lifetime value is nearly identical across all groups, hovering between 5,300 and 6,300 regardless of education or gender.
- Highest Spenders: The biggest individual values appear in High School or Below (Male) at 83,325 and Bachelor (Female) at 73,226.
- Advanced Degrees Cap Out Lower: Customers with a Master or Doctor degree reach much lower maximums, peaking between roughly 32,600 and 51,000.
- Equal Baseline: Minimum values stay uniform across every tier at around 1,900 to 2,400.
- Main Takeaway: Higher academic education does not lead to higher customer lifetime value.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here